tiledb/sm/filter/webp_filter.cc

```diff
@@ -0,0 +1,324 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements class WebpFilter.
+ */
+
+#ifdef TILEDB_WEBP
+
+#include "tiledb/sm/filter/webp_filter.h"
+#include "tiledb/common/status.h"
+#include "tiledb/sm/tile/tile.h"
+
+#include "webp/decode.h"
+#include "webp/encode.h"
+
+using namespace tiledb::common;
+
+namespace tiledb::sm {
+void WebpFilter::dump(FILE* out) const {
+  if (out == nullptr)
+    out = stdout;
+  fprintf(out, "WebpFilter");
+}
+
+Status WebpFilter::run_forward(
+    const Tile& tile,
+    Tile* const,
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  if (tile.type() != Datatype::UINT8) {
+    throw StatusException(Status_FilterError("Unsupported input type"));
+  }
+  return run_forward(input_metadata, input, output_metadata, output);
+}
+
+Status WebpFilter::run_forward(
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  auto input_parts = input->buffers();
+  auto num_parts = static_cast<uint32_t>(input_parts.size());
+  uint32_t metadata_size = sizeof(uint32_t) + (num_parts * sizeof(uint32_t));
+  throw_if_not_ok(output_metadata->append_view(input_metadata));
+  throw_if_not_ok(output_metadata->prepend_buffer(metadata_size));
+  throw_if_not_ok(output_metadata->write(&num_parts, sizeof(uint32_t)));
+
+  int extent_y = extents_.first, extent_x = extents_.second,
+      pixel_depth = format_ < WebpInputFormat::WEBP_RGBA ? 3 : 4;
+  // X should be divisible by colorspace pixel_depth or RGB values will skew.
+  if (extent_x % pixel_depth != 0) {
+    throw StatusException(Status_FilterError(
+        pixel_depth == 3 ?
+            "Colorspace with no alpha must use extents divisible by 3" :
+            "Colorspace with alpha must use extents divisible by 4"));
+  }
+
+  for (const auto& i : input_parts) {
+    auto data = static_cast<const uint8_t*>(i.data());
+    // Number of bytes encoded; Encoded result data buffer.
+    size_t enc_size = 0;
+    uint8_t* result = nullptr;
+    // Cleanup allocated data when we leave scope.
+    ScopedExecutor cleanup([&]() {
+      if (enc_size > 0 && result != nullptr)
+        WebPFree(result);
+    });
+    // We divide extent_x by colorspace value count to get pixel-width of image,
+    // extent_x currently represents row stride.
+    switch (format_) {
+      case WebpInputFormat::WEBP_RGB:
+        if (lossless_) {
+          enc_size = WebPEncodeLosslessRGB(
+              data, extent_x / pixel_depth, extent_y, extent_x, &result);
```

Comment:
Thanks for the review! I have, but the test cases I had pushed here were using very even numbers so I updated them last night to test against more odd shaped images.

I think because this filter is for dense arrays the data size will always be the size of the max tile chunk, and the excess data would have been zeroed in [dense_tiler.cc](https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/query/writers/dense_tiler.cc#L285-L288). So for these boundary tiles webp might compress some excess zeros if the tiles don't align perfectly with the image dimensions. I checked this in a debugger using a 20x20 image with extents of 15 and 30 for Y, X dimensions. In this case things 'just worked', but I'll think about this a bit and see if I can figure a way to strip this excess data, I do think it would be an improvement and we would see a storage benefit in the size of the ingested array.